src/Simulation/Common/IQuantumProcessor.cs

```diff
@@ -391,6 +391,41 @@ public interface IQuantumProcessor
         /// <param name="qubits">Register to apply the exponent to.</param>
         void ControlledExpFrac(IQArray<Qubit> controls, IQArray<Pauli> paulis, long numerator, long power, IQArray<Qubit> qubits);
 
+        /// <summary>
+        /// Used as extensibility mechanism to add custom operations to IQuantumProcessor. 
+        /// </summary>
+        /// <remarks>
+        /// The custom operation may use only some of the parameters. Depending on the nature of the operation, all of them may be optional. 
+        /// A variety of parameter types are provided for flexibility.
+        /// </remarks>
+        /// <param name="operationId">The id of the custom operation. It is up to the custom operation implementor to define the meaning.</param>
+        /// <param name="qubits">Array of qubits to apply the operation to. Their meaning is specific to the custom operation.</param>
+        /// <param name="paulis">Array of single-qubit Pauli values. Their meaning is specific to the custom operation.</param>
+        /// <param name="longs">Array of long parameters. Their meaning is specific to the custom operation.</param>
+        /// <param name="doubles">Array of double parameters. Their meaning is specific to the custom operation.</param>
+        /// <param name="bools">Array of bool parameters. Their meaning is specific to the custom operation.</param>
+        /// <param name="results">Array of Result parameters. Their meaning is specific to the custom operation.</param>
+        /// <param name="strings">Array of string parameters. Their meaning is specific to the custom operation.</param>
+        void CustomOperation(long operationId, IQArray<Qubit> qubits, IQArray<Pauli> paulis, IQArray<long> longs, IQArray<double> doubles, IQArray<bool> bools, IQArray<Result> results, IQArray<string> strings);
```

Comment:
>CustomOperation [](start = 13, length = 15)

How will we decide which gates are explicitly spelled in the interface and which go through the custom operation?

Reply:
The set of gates that are explicitly spelled out is fixed in the definition of Q# intrinsics. I believe this set was decided based on being a minimally reasonable set supporting a closure of quantum operations and supporting functionality. But anyway, it is decided and users cannot change these. Custom operation allows users to add any gates they deem necessary for them. Usually for more complex gates for more efficient execution, and also for custom supporting functionality that is supported only by their custom quantum processor. 

---
In reply to: [413475308](https://github.com/microsoft/qsharp-runtime/pull/172#discussion_r413475308) [](ancestors = 413475308)

Reply:
Any custom operation can always be defined as intrinsic. The problem to solve is how to make it more user friendly to do that. There are several possibilities on how to achieve that, and we should discuss these.